compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -867,6 +874,50 @@ impl InitMask {
     }
 }
 
+/// Yields [`InitChunk`]s. See [`InitMask::range_as_init_chunks`].
+pub struct InitChunkIter<'a> {
+    init_mask: &'a InitMask,
+    /// The current byte index into `init_mask`.
+    start: Size,
+    /// The end byte index into `init_mask`.
+    end: Size,
+}
+
+/// A contiguous chunk of initialized or uninitialized memory.
+pub enum InitChunk {
+    Init(Range<Size>),
+    Uninit(Range<Size>),
+}
+
+impl<'a> InitChunkIter<'a> {
+    fn new(init_mask: &'a InitMask, start: Size, end: Size) -> Self {
+        assert!(start <= end);
+        assert!(end <= init_mask.len);
+        Self { init_mask, start, end }
+    }
+}
+
+impl<'a> Iterator for InitChunkIter<'a> {
+    type Item = InitChunk;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.start >= self.end {
+            return None;
+        }
+
+        let is_init = self.init_mask.get(self.start);
+        let end_of_chunk = (self.start.bytes()..self.end.bytes())
+            .map(Size::from_bytes)
+            .find(|&i| self.init_mask.get(i) != is_init)
```

Comment:
I think we could improve on this by checking entire blocks instead of going through the entries in the block, but again, doesn't have to happen in this PR (please leave a FIXME comment for that though).

Reply:
Dealing with the {start,end} of the {search range,initialized chunk} being unaligned with block boundaries seems subtle (and very branchy in a naive implementation) so I'll just copy the fixme from the other place this was done